### PR TITLE
added base_files for clip grad

### LINKED ISSUE #50: Dealing with model collapse
**Is your feature request related to a problem? Please describe.**
Sometimes, because of the way a model gets initialized/trained, *model collapse* (i.e., loss and weights all become `nan`) can happen.

**Describe the solution you'd like**
Instead of having re-initialize the complete training sequence, we should explore ways to make GaNDLF automatically do this.

**Describe alternatives you've considered**
Open to other suggestions.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by Geeks-Sid**
Fixes #50 

## Proposed Changes

- This adds gradient clipping by normalizing gradients by norm, by value and adaptive gradient clipping.

## Checklist

- [x] I have read the [`CONTRIBUTING`](https://github.com/CBICA/GaNDLF/blob/master/CONTRIBUTING.md) guide
- [x] My PR is based from the [current GaNDLF master ](https://garygregory.wordpress.com/2016/11/10/how-to-catch-up-my-git-fork-to-master/)
- [x] Non-breaking change (would not break existing functionality): if changes breaks current code, please provide as many details as possible
- [ ] Function/class source code documentation added/updated
- [x] Code has been [blacked](https://github.com/psf/black#usage) for style consistency
- [x] If applicable, version information [has been updated in GANDLF/__init__.py](../GANDLF/__init__.py)
- [ ] If adding a submodule, add to list of expections for black styling in [pyproject.toml](../pyproject.toml) file 
- [x] [Usage documentation](https://github.com/CBICA/GaNDLF/blob/master/docs) has been updated, if appropriate
- [ ] [History](https://github.com/CBICA/GaNDLF/blob/master/HISTORY.md) has been updated, if appropriate
- [x] Tests added or modified to cover the changes
- [x] Integration tests passed locally by running `pytest`

### CODE REVIEW

File: GANDLF/preprocessing.py [hunk 1]
```diff
@@ -133,17 +133,11 @@ def __init__(self, masking_method: TypeMaskingMethod = None, **kwargs):
         self.args_names = ("masking_method",)
 
     def apply_normalization(
-        self,
-        subject: Subject,
-        image_name: str,
-        mask: torch.Tensor,
+        self, subject: Subject, image_name: str, mask: torch.Tensor,
```
Comment (sarthakpati) [inline GANDLF/preprocessing.py:136]: This file isn't blacked.

**Comment by sarthakpati** [comment]
This code hasn't been blacked, BTW. Mulitple inconsistencies found in the styling.

**Comment by sarthakpati** [comment]
Drafted till #76 is merged.

File: GANDLF/misc_utils/adaptive_gradient_clipping.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    else:
+        # This should work for nn.ConvNd (where N is the number of dimensions)
+        # and nn.Linear where number of output dimensions is first
+        # in the weight/kernel tensor passed.
+        # This might need something for a few unplanned scenarios
+        # but we dont have to deal with them here.
+        return x.norm(norm_type, dim=tuple(range(1, x.ndim)), keepdim=True)
+
+
+def adaptive_gradient_clip(parameters, clip_factor=0.01, eps=1e-3, norm_type=2.0):
```
Comment (sarthakpati) [inline GANDLF/misc_utils/adaptive_gradient_clipping.py:21]:
```suggestion
def adaptive_gradient_clip_(parameters, clip_factor=0.01, eps=1e-3, norm_type=2.0):
```

**Review by sarthakpati — APPROVED** [review]
LGTM, needs a few changes that we can do after the merge.
